rfc/rfc-60/rfc-60.md

```diff
@@ -84,23 +119,85 @@ public interface HoodieStorageStrategy extends Serializable {
   /**
    * Return a storage location for the given filename.
    *
-   * @param fileId data file ID
+   * @param path fileName
    * @return a storage location string for a data file
    */
-  String storageLocation(String fileId);
+  StoragePath storageLocation(String path, String instantTime);
 
   /**
-   * Return a storage location for the given partition and filename.
+   * Return all possible StoragePaths
    *
-   * @param partitionPath partition path for the file
-   * @param fileId data file ID
-   * @return a storage location string for a data file
+   * @param partitionPath
+   * @param checkExist check if StoragePath is truly existed or not. 
+   * @return a st of storage partition path
+   */
+  Set<StoragePath> getAllLocations(String partitionPath, boolean checkExist);
```

Comment:
Where can this API be useful? 

I think we need to think twice about it. Adding this API means Hudi should be able to infer relative partition locations solely based on partition path. And for `Object Store Strategy` this means all the data under the same partition would be stored under the same physical location again, because we can only calculate the hash based on two arguments that got passed in.

Reply:
This is needed, for example, in the `AbstractTableFileSystemView#ensurePartitionLoadedCorrectly(String partition)` method. 

An external relative path partition is passed in, and currently, the base path is directly prefixed to it before performing the listPartition(Path partitionPath) operation. 

For Hudi tables with the Federated Storage Layout feature enabled, the same relative path might map to multiple physical paths. Here, the HoodieStorageStrategy#getAllLocations method is required.

Reply:
For `HoodieObjectStoreStrategy`, can we design it such that `HoodieObjectStoreStrategy` specifies `hoodie.storage.path` as `s3://<table_storage_bucket>/hudi_location/`, and then all data is written under `s3://<table_storage_bucket>/hudi_location/` like
```
s3://<table_storage_bucket>/hudi_location/0bfb3d6e/<hudi_table_name>/9320889c-8537-4aa7-a63e-ef088b9a21ce-0_9-11-51_20220301005056692.parquet
s3://<table_storage_bucket>/hudi_location/0bfb3d6e/<hudi_table_name>/a62aa56b-d55e-4a2b-88a6-d603ef26775c-0_8-11-50_20220301005056692.parquet
s3://<table_storage_bucket>/hudi_location/0bfb3d6e/<hudi_table_name>/.4b0c6b40-2ac0-4a1c-a26f-6338aa4db22e-0_6-11-48_20220301005056692.log.1_0-22-26
s3://<table_storage_bucket>/hudi_location/0bfb3d6e/<hudi_table_name>/.075f3295-def8-4a42-a927-07fd2dd2976c-0_7-11-49_20220301005056692.log.1_0-22-26
```
In the `HoodieObjectStoreStrategy#getAllLocations()` method, first list the `hoodie.storage.path`, and then append the relative path.

getAllLocations() is not in `hot path`, so list cost is not very expensive.

Reply:
This API was changed in #12460, could you keep this updated as well?